src/test/java/org/elasticsearch/bootstrap/SecurityTests.java

```diff
@@ -19,30 +19,77 @@
+        Path policyFile = Security.processTemplate(new ByteArrayInputStream(new byte[0]), environment);
+      
+        ProtectionDomain domain = getClass().getProtectionDomain();
+        Policy policy = Policy.getInstance("JavaPolicy", new URIParameter(policyFile.toUri()));
+        // the fake es home
+        assertTrue(policy.implies(domain, new FilePermission(esHome.toString(), "read")));
+        // its parent
+        assertFalse(policy.implies(domain, new FilePermission(path.toString(), "read")));
+        // some other sibling
+        assertFalse(policy.implies(domain, new FilePermission(path.resolve("other").toString(), "read")));
     }
-    
-    /** test template processing */
-    public void testTemplateProcessing() throws Exception {
+
+    /** test generated permissions for all configured paths */
+    public void testEnvironmentPaths() throws Exception {
         Path path = createTempDir();
-        
-        byte results[] = Security.createPermissions(Collections.singleton(path));
-        String unicode = new String(results, StandardCharsets.UTF_8);
-        // try not to make this test too fragile or useless
-        assertTrue(unicode.contains("grant {"));
-        assertTrue(unicode.contains(Security.encode(path)));
-        assertTrue(unicode.contains("read"));
-        assertTrue(unicode.contains("write"));
+
+        ImmutableSettings.Builder settingsBuilder = ImmutableSettings.builder();
+        settingsBuilder.put("path.home", path.resolve("home").toString());
+        settingsBuilder.put("path.conf", path.resolve("conf").toString());
+        settingsBuilder.put("path.plugins", path.resolve("plugins").toString());
+        settingsBuilder.putArray("path.data", path.resolve("data1").toString(), path.resolve("data2").toString());
+        settingsBuilder.put("path.logs", path.resolve("logs").toString());
+        Settings settings = settingsBuilder.build();
+
+        Environment environment = new Environment(settings);        
+        Path policyFile = Security.processTemplate(new ByteArrayInputStream(new byte[0]), environment);
+     
+        ProtectionDomain domain = getClass().getProtectionDomain();
+        Policy policy = Policy.getInstance("JavaPolicy", new URIParameter(policyFile.toUri()));
+
+        // check that all directories got permissions:
+        // homefile: this is needed unless we break out rules for "lib" dir.
+        // TODO: make read-only
+        assertTrue(policy.implies(domain, new FilePermission(environment.homeFile().toString(), "read,readlink,write,delete")));
+        // config file
+        // TODO: make read-only
+        assertTrue(policy.implies(domain, new FilePermission(environment.configFile().toString(), "read,readlink,write,delete")));
+        // plugins: r/w, TODO: can this be minimized?
+        assertTrue(policy.implies(domain, new FilePermission(environment.pluginsFile().toString(), "read,readlink,write,delete")));
```

Comment:
well, it wasnt clear to me, because our securitymanager in tests gives r/w permissions to the entire temp directories allocated to it. So I really want to be able to test that read-only changes work by applying these permissions in our tests, too. For this to work I think we need a custom SM subclass that lets you add "nodes" to it. That's why i put a TODO and still give r/w permissions everywhere, because its consistent with the way we currently test.

As far as configuration, personally I think we should keep these options to a minimum and try to come up with an out-of-box configuration that is robust and secure for 99% of users. I imagine SM changes will break some plugins but we have to change things to make progress. And users can always disable SM completely if there are problems.